### PR TITLE
[SYCL] [CUDA]Add experimental cuda interop

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR adds an experimental CUDA-interop that reflects the proposed CUDA backend spec in https://github.com/KhronosGroup/SYCL-Docs/pull/197. The changes work with the CUDA CTS interop checks https://github.com/KhronosGroup/SYCL-CTS/pull/336. 

The changes are kept experimental so as to not break oneMKL, and oneDNN which use CUDA interops.

The experimental interop is enabled by defining the `SYCL_EXT_ONEAPI_BACKEND_CUDA_EXPERIMENTAL` macro and including the header file `sycl/ext/oneapi/experimental/backend/cuda.hpp`.

### CODE REVIEW

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -2025,12 +2116,38 @@ pi_result cuda_piextContextGetNativeHandle(pi_context context,
...
-      "Creation of PI context from native handle not implemented");
-  return {};
+pi_result cuda_piextContextCreateWithNativeHandle(pi_native_handle nativeHandle,
+                                                  pi_uint32 num_devices,
+                                                  const pi_device *devices,
+                                                  bool ownNativeHandle,
+                                                  pi_context *piContext) {
+  (void)num_devices;
+  (void)devices;
+  (void)ownNativeHandle;
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/pi_cuda.cpp:2126]: don't you want to check (also in other places) that unused arguments are either unset, or set to reasonable values?

**Comment by smaslov-intel** [comment]
The changes look good high-level, but can't you split them into multiple PRs?
Also, are you adding E2E tests to check the interop?

**Comment by smaslov-intel** [comment]
> Also, are you adding E2E tests to check the interop?

I see you are adding tests in https://github.com/KhronosGroup/SYCL-CTS/pull/336
When/how are they being run?

**Comment by smaslov-intel** [comment]
> The experimental interop is enabled by defining the SYCL_EXT_ONEAPI_BACKEND_CUDA_EXPERIMENTAL macro 

What does "experimental" denotes here and when can/will it become non-experimental?

**Comment by AidanBeltonS** [comment]
> The changes look good high-level, but can't you split them into multiple PRs? Also, are you adding E2E tests to check the interop?

Sure, I think this can be split into a few PRs. It might be a bit tricky as there may be some dependencies between the different PRs. Would you prefer this to be separated into a few submissions?
I can add some E2E tests to the llvm-test-suite, as well.


> I see you are adding tests in [KhronosGroup/SYCL-CTS#336](https://github.com/KhronosGroup/SYCL-CTS/pull/336)
> When/how are they being run?

I am running the CUDA CTS tests locally to verify this PR passes the SYCL-CTS. I run them when I make changes to either this PR or the CUDA CTS tests.


> What does "experimental" denotes here and when can/will it become non-experimental?

This is experimental as there is still some discussion regarding the CUDA backend spec. In addition both oneMKL and oneDNN use the exiting interop, if these changes were to overwrite the current interop behaviour both of these libraries would break. 
I think this can be non-experimental when the CUDA backend spec is finalised and agreed, and oneMKL and oneDNN are updated to use this interop.

**Comment by AidanBeltonS** [comment]
> Yes, please. Even splitting to 2 will make it more manageable.

Sure I will work on splitting this up.



> > 
> 
> Please clarify what is the "current interop"? I am not seeing any interop support in the current CUDA Plugin.

Although there is no support in the CUDA plugin for creating SYCL objects with native types, there is functionality to return native types from SYCL objects. 

The returned native types are being changed by the CUDA backend spec, for example `CUdeviceptr` is being replaced with `DataT *`. So by current I am referring to the input and return native types defined in `include/CL/sycl/detail/backend_traits_cuda.hpp` compared to the experimental native types defined in `/include/sycl/ext/oneapi/experimental/backend/backend_traits_cuda.hpp`.

**Comment by smaslov-intel** [comment]
> Sure I will work on splitting this up.

Thank you!

> The returned native types are being changed by the CUDA backend spec, for example CUdeviceptr is being replaced with DataT *. So by current I am referring to the input and return native types defined in include/CL/sycl/detail/backend_traits_cuda.hpp compared to the experimental native types defined in /include/sycl/ext/oneapi/experimental/backend/backend_traits_cuda.hpp.

I see. I think it would still be better to have a plan from the beginning how the new interop becomes the default and the only one eventually. And, then, how to get it out of the "experimental" envelope. @gmlueck : would you have any suggestions here?

**Comment by steffenlarsen** [comment]
> I see. I think it would still be better to have a plan from the beginning how the new interop becomes the default and the only one eventually. And, then, how to get it out of the "experimental" envelope.

Since this is an ABI change we cannot reasonably have a deprecation period for the change, so I don't personally see any other option than promoting it out of experimental (replacing the old API) once we are in an ABI break window. Preferably this would happen when the backend specification is deemed fully ready and is implemented in this experimental guard.

**Comment by smaslov-intel** [comment]
> > I see. I think it would still be better to have a plan from the beginning how the new interop becomes the default and the only one eventually. And, then, how to get it out of the "experimental" envelope.
> 
> Since this is an ABI change we cannot reasonably have a deprecation period for the change, so I don't personally see any other option than promoting it out of experimental (replacing the old API) once we are in an ABI break window. Preferably this would happen when the backend specification is deemed fully ready and is implemented in this experimental guard.

This sounds right. We need to keep track of this change (remove "experimental") needed when ABI break is allowed. Shouldn't we add deprecation messages on the "current" interop now?

**Comment by steffenlarsen** [comment]
> Shouldn't we add deprecation messages on the "current" interop now?

Since the switch is guarded by `SYCL_EXT_ONEAPI_BACKEND_CUDA_EXPERIMENTAL` I am not sure if we want to deprecate the current API as it would have to tell the user to define `SYCL_EXT_ONEAPI_BACKEND_CUDA_EXPERIMENTAL` in order for the warnings to go away. Most often we can't issue deprecation warnings prior to ABI break as we can't always offer a deprecation period, though I suppose we _could_ do it in this case.

**Comment by smaslov-intel** [comment]
We can only remove old interface after it was deprecated for some time. I think it is OK to tell users that we are moving to the new version, which is currently implemented under `SYCL_EXT_ONEAPI_BACKEND_CUDA_EXPERIMENTAL` and they should start coding with it, because it will become the default (and the only one) in the next major update or so.

**Comment by AerialMantis** [comment]
I'd like to keep this pull request open, we've not been very active on it recently, but we plan to come back to it again soon.

**Pull request closed without merging**